data/builder/dirshard.go

```diff
@@ -0,0 +1,176 @@
+package builder
+
+import (
+	"fmt"
+
+	bitfield "github.com/ipfs/go-bitfield"
+	"github.com/ipfs/go-unixfsnode/data"
+	dagpb "github.com/ipld/go-codec-dagpb"
+	"github.com/ipld/go-ipld-prime"
+	"github.com/multiformats/go-multihash"
+)
+
+type shard struct {
+	// metadata about the shard
+	hasher uint64
+	size   int
+	width  int
+	depth  int
+
+	children map[int]entry
+}
+
+// a shard entry is either another shard, or a direct link.
+type entry struct {
+	*shard
+	*hamtLink
+}
+
+// a hamtLink is a member of the hamt - the file/directory pointed to, but
+// stored with it's hashed key used for addressing.
+type hamtLink struct {
+	hash hashBits
+	dagpb.PBLink
+}
+
+// BuildUnixFSShardedDirectory will build a hamt of unixfs hamt shards encoing a directory with more entries
+// than is typically allowed to fit in a standard IPFS single-block unixFS directory.
+func BuildUnixFSShardedDirectory(size int, hasher uint64, entries []dagpb.PBLink, ls *ipld.LinkSystem) (ipld.Link, error) {
+	// hash the entries
+	h, err := multihash.GetHasher(hasher)
+	if err != nil {
+		return nil, err
+	}
+	he := make([]hamtLink, 0, len(entries))
+	for _, e := range entries {
+		name := e.Name.Must().String()
+		sum := h.Sum([]byte(name))
+		he = append(he, hamtLink{
+			sum,
+			e,
+		})
+	}
+
+	sharder := shard{
+		hasher: hasher,
+		size:   size,
+		width:  len(fmt.Sprintf("%X", size-1)),
+		depth:  0,
+
+		children: make(map[int]entry),
+	}
+
+	for _, e := range he {
+		sharder.add(e)
+	}
+
+	return sharder.serialize(ls)
+}
+
+func (s *shard) add(lnk hamtLink) error {
+	// get the bucket for lnk
+	bucket, err := lnk.hash.Slice(s.depth*s.size, s.size)
```

Comment:
I'm pretty sure you're missing a log2 somewhere here. There's a lot of calculation of log2 in the HAMT sharding code, and it's cause even thought we want 256 width, that only means 8-bits per item.

there are tests in the go-unixfs implementation -- how hard would it be to substantially copy them to very you have the same result?

Reply:
maybe I'm totally wrong -- that's part of why I'm wondering if we have a way to equivalence test this